File: ios-app/BayitPlusApp/ViewModels/FriendsViewModel.swift
import BayitCore
import Foundation
import Observation

/// ViewModel for the Friends screen -- manages friend list, requests, and user search.
@MainActor
@Observable
final class FriendsViewModel {
    private(set) var friends: [Friend] = []
    private(set) var incomingRequests: [FriendRequest] = []
    private(set) var outgoingRequests: [FriendRequest] = []
    private(set) var searchResults: [UserSearchResult] = []
    private(set) var isLoading = false
    private(set) var isSearching = false
    private(set) var error: String?

    var searchQuery = ""

    private let repository: any FriendsRepository
    private let logger = BayitLogger(category: "Friends")
    private let searchLimit = 20

    init(repository: any FriendsRepository) {
        self.repository = repository
    }

    // MARK: - Load Data

    @MainActor
    func loadFriends() async {
        isLoading = true
        error = nil

        do {
            friends = try await repository.fetchFriends()
            logger.info("Friends loaded", context: ["count": String(friends.count)])
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to load friends", error: error)
        }

        isLoading = false
    }

    @MainActor
    func loadRequests() async {
        do {
            let response = try await repository.fetchRequests()
            incomingRequests = response.incoming
            outgoingRequests = response.outgoing
            logger.info("Requests loaded", context: [
                "incoming": String(response.incoming.count),
                "outgoing": String(response.outgoing.count)
            ])
        } catch {
            logger.error("Failed to load requests", error: error)
        }
    }

    @MainActor
    func loadAll() async {
        isLoading = true
        error = nil

        async let friendsResult: Void = loadFriendsOnly()
        async let requestsResult: Void = loadRequestsOnly()
        _ = await (friendsResult, requestsResult)

        isLoading = false
    }

    // MARK: - Actions

    @MainActor
    func sendRequest(to receiverId: String) async {
        do {
            try await repository.sendRequest(receiverId: receiverId)
            logger.info("Friend request sent", context: ["receiverId": receiverId])
            await loadRequests()
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to send friend request", error: error)
        }
    }

    @MainActor
    func acceptRequest(_ requestId: String) async {
        do {
            try await repository.acceptRequest(requestId: requestId)
            logger.info("Request accepted", context: ["requestId": requestId])
            await loadAll()
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to accept request", error: error)
        }
    }

    @MainActor
    func rejectRequest(_ requestId: String) async {
        do {
            try await repository.rejectRequest(requestId: requestId)
            logger.info("Request rejected", context: ["requestId": requestId])
            incomingRequests.removeAll { $0.id == requestId }
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to reject request", error: error)
        }
    }

    @MainActor
    func searchUsers() async {
        let trimmed = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            searchResults = []
            return
        }

        isSearching = true
        do {
            searchResults = try await repository.searchUsers(query: trimmed, limit: searchLimit)
            logger.info("User search completed", context: [
                "query": trimmed, "resultCount": String(searchResults.count)
            ])
        } catch {
            logger.error("User search failed", error: error)
        }
        isSearching = false
    }

    // MARK: - Private Helpers

    @MainActor
    private func loadFriendsOnly() async {
        do {
            friends = try await repository.fetchFriends()
        } catch {
            self.error = error.localizedDescription
            logger.error("Failed to load friends", error: error)
        }
    }

    @MainActor
    private func loadRequestsOnly() async {
        do {
            let response = try await repository.fetchRequests()
            incomingRequests = response.incoming
            outgoingRequests = response.outgoing
        } catch {
            logger.error("Failed to load requests", error: error)
        }
    }
}

File: android-app/feature/feature-social/src/main/java/tv/bayit/plus/feature/social/friends/FriendsViewModel.kt
package tv.bayit.plus.feature.social.friends

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tv.bayit.plus.core.common.BayitResult
import tv.bayit.plus.core.common.logging.BayitLogger
import tv.bayit.plus.core.data.repository.FriendsRepository
import tv.bayit.plus.core.model.Friend
import tv.bayit.plus.core.model.FriendRequest
import javax.inject.Inject

@HiltViewModel
class FriendsViewModel @Inject constructor(
    private val friendsRepository: FriendsRepository,
    private val logger: BayitLogger,
) : ViewModel() {

    private val _uiState = MutableStateFlow<FriendsUiState>(FriendsUiState.Loading)
    val uiState: StateFlow<FriendsUiState> = _uiState.asStateFlow()

    private val _selectedTab = MutableStateFlow(FriendsTab.FRIENDS)
    val selectedTab: StateFlow<FriendsTab> = _selectedTab.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    init {
        loadFriends()
    }

    fun selectTab(tab: FriendsTab) {
        _selectedTab.value = tab
        when (tab) {
            FriendsTab.FRIENDS -> loadFriends()
            FriendsTab.PENDING -> loadPendingRequests()
            FriendsTab.SEARCH -> Unit
        }
    }

    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
    }

    fun searchUsers() {
        val query = _searchQuery.value
        if (query.isBlank()) return
        viewModelScope.launch {
            _uiState.value = FriendsUiState.Loading
            logger.debug("Searching users", mapOf("query" to query))
            when (val result = friendsRepository.searchUsers(query)) {
                is BayitResult.Success -> {
                    val users = result.data.filterIsInstance<Friend>()
                    _uiState.value = FriendsUiState.SearchResults(users)
                }
                is BayitResult.Error -> {
                    logger.error("User search failed", result.exception)
                    _uiState.value = FriendsUiState.Error(
                        result.message ?: result.exception.message.orEmpty(),
                    )
                }
                is BayitResult.Loading -> Unit
            }
        }
    }

    fun sendRequest(userId: String) {
        viewModelScope.launch {
            logger.info("Sending friend request", mapOf("userId" to userId))
            when (val result = friendsRepository.sendRequest(userId)) {
                is BayitResult.Success -> logger.info("Friend request sent")
                is BayitResult.Error -> logger.error("Friend request failed", result.exception)
                is BayitResult.Loading -> Unit
            }
        }
    }

    fun acceptRequest(requestId: String) {
        viewModelScope.launch {
            logger.info("Accepting friend request", mapOf("requestId" to requestId))
            when (val result = friendsRepository.acceptRequest(requestId)) {
                is BayitResult.Success -> loadPendingRequests()
                is BayitResult.Error -> logger.error("Accept request failed", result.exception)
                is BayitResult.Loading -> Unit
            }
        }
    }

    fun declineRequest(requestId: String) {
        viewModelScope.launch {
            logger.info("Declining friend request", mapOf("requestId" to requestId))
            when (val result = friendsRepository.declineRequest(requestId)) {
                is BayitResult.Success -> loadPendingRequests()
                is BayitResult.Error -> logger.error("Decline request failed", result.exception)
                is BayitResult.Loading -> Unit
            }
        }
    }

    private fun loadFriends() {
        viewModelScope.launch {
            _uiState.value = FriendsUiState.Loading
            logger.debug("Loading friends list")
            when (val result = friendsRepository.getFriends()) {
                is BayitResult.Success -> {
                    val friends = result.data.filterIsInstance<Friend>()
                    logger.info("Friends loaded", mapOf("count" to friends.size.toString()))
                    _uiState.value = FriendsUiState.FriendsList(friends)
                }
                is BayitResult.Error -> {
                    logger.error("Friends load failed", result.exception)
                    _uiState.value = FriendsUiState.Error(
                        result.message ?: result.exception.message.orEmpty(),
                    )
                }
                is BayitResult.Loading -> Unit
            }
        }
    }

    private fun loadPendingRequests() {
        viewModelScope.launch {
            _uiState.value = FriendsUiState.Loading
            logger.debug("Loading pending requests")
            when (val result = friendsRepository.getPendingRequests()) {
                is BayitResult.Success -> {
                    val requests = result.data.filterIsInstance<FriendRequest>()
                    _uiState.value = FriendsUiState.PendingRequests(requests)
                }
                is BayitResult.Error -> {
                    logger.error("Pending requests load failed", result.exception)
                    _uiState.value = FriendsUiState.Error(
                        result.message ?: result.exception.message.orEmpty(),
                    )
                }
                is BayitResult.Loading -> Unit
            }
        }
    }
}

sealed interface FriendsUiState {
    data object Loading : FriendsUiState
    data class FriendsList(val friends: List<Friend>) : FriendsUiState
    data class PendingRequests(val requests: List<FriendRequest>) : FriendsUiState
    data class SearchResults(val users: List<Friend>) : FriendsUiState
    data class Error(val message: String) : FriendsUiState
}

enum class FriendsTab { FRIENDS, PENDING, SEARCH }
